
<img src="Datasets\Tp1.png">





<h1 align="center">FEKIH Firas</h1>
<h1 align="center">3 DNI 2</h1>


![We Can Use Giphy & Emojis Too](http://www.minddirect.in/wp-content/uploads/2020/10/data-cleaning.gif)



### Importer les bibliothèques **numpy** et **pandas** et importer functools de reduce


In [1]:
# Importer les bibliothèques nécessaires

import pandas as pd
import numpy as np
from functools import reduce

# PART I

Explorer la dataset `BL-Flickr-Images-Book.csv` 

**NB:** faites attentions au chemin

## 1 Exploration du dataset

In [2]:
# Exporter BL-Flickr-Images-Book.csv dans un data frame nomé df
df = pd.read_csv('Datasets/BL-Flickr-Images-Book.csv')
# afficher les 5 premières lignes
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [3]:
# Exporter BL-Flickr-Images-Book.csv dans un data frame nomé df
df = pd.read_csv('Datasets/BL-Flickr-Images-Book.csv')
# afficher les 5 dernières lignes
df.tail()

Identifier Edition Statement Place of Publication Date of Publication  \
8282     4158088               NaN               London                1838   
8283     4158128               NaN                Derby            1831, 32   
8284     4159563               NaN               London           [1806]-22   
8285     4159587               NaN  Newcastle upon Tyne                1834   
8286     4160339               NaN               London             1834-43   

                    Publisher  \
8282                      NaN   
8283          M. Mozley & Son   
8284  T. Cadell and W. Davies   
8285         Mackenzie & Dent   
8286                      NaN   

                                                  Title  \
8282  The Parochial History of Cornwall, founded on,...   
8283  The History and Gazetteer of the County of Der...   
8284  Magna Britannia; being a concise topographical...   
8285  An historical, topographical and descriptive v...   
8286  Collectanea Topographica et Genealogica. [Firs...   

                                                 Author  \
8282                 GIDDY, afterwards GILBERT, Davies.   
8283                         GLOVER, Stephen - of Derby   
8284  LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
8285                              Mackenzie, E. (Eneas)   
8286                                                NaN   

                                           Contributors  Corporate Author  \
8282  BOASE, Henry Samuel.|HALS, William.|LYSONS, Da...               NaN   
8283                                     NOBLE, Thomas.               NaN   
8284           GREGSON, Matthew.|LYSONS, Samuel - F.R.S               NaN   
8285                               ROSS, M. - of Durham               NaN   
8286            BANDINEL, Bulkeley.|Nichols, John Gough               NaN   

      Corporate Contributors Former owner  Engraver Issuance type  \
8282                     NaN          NaN       NaN   monographic   
8283                     NaN          NaN       NaN   monographic   
8284                     NaN          NaN       NaN   monographic   
8285                     NaN          NaN       NaN   monographic   
8286                     NaN          NaN       NaN    continuing   

                                             Flickr URL  \
8282  http://www.flickr.com/photos/britishlibrary/ta...   
8283  http://www.flickr.com/photos/britishlibrary/ta...   
8284  http://www.flickr.com/photos/britishlibrary/ta...   
8285  http://www.flickr.com/photos/britishlibrary/ta...   
8286  http://www.flickr.com/photos/britishlibrary/ta...   

                                             Shelfmarks  
8282  British Library HMNTS|British Library HMNTS 10...  
8283  British Library HMNTS|British Library HMNTS 10...  
8284  British Library HMNTS|British Library HMNTS 19...  
8285  British Library HMNTS|British Library HMNTS 10...  
8286  British Library HMNTS|British Library HMNTS 79...

In [4]:
df.columns

Index(['Identifier', 'Edition Statement', 'Place of Publication',
       'Date of Publication', 'Publisher', 'Title', 'Author', 'Contributors',
       'Corporate Author', 'Corporate Contributors', 'Former owner',
       'Engraver', 'Issuance type', 'Flickr URL', 'Shelfmarks'],
      dtype='object')

In [5]:
df['Shelfmarks']

0                       British Library HMNTS 12641.b.30.
1                       British Library HMNTS 12626.cc.2.
2                       British Library HMNTS 12625.dd.1.
3                     British Library HMNTS 10369.bbb.15.
4                        British Library HMNTS 9007.d.28.
                              ...                        
8282    British Library HMNTS|British Library HMNTS 10...
8283    British Library HMNTS|British Library HMNTS 10...
8284    British Library HMNTS|British Library HMNTS 19...
8285    British Library HMNTS|British Library HMNTS 10...
8286    British Library HMNTS|British Library HMNTS 79...
Name: Shelfmarks, Length: 8287, dtype: object

## 2 Jeter un coup d'oeil dans le dataset

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8287 entries, 0 to 8286
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Identifier              8287 non-null   int64  
 1   Edition Statement       773 non-null    object 
 2   Place of Publication    8287 non-null   object 
 3   Date of Publication     8106 non-null   object 
 4   Publisher               4092 non-null   object 
 5   Title                   8287 non-null   object 
 6   Author                  6509 non-null   object 
 7   Contributors            8287 non-null   object 
 8   Corporate Author        0 non-null      float64
 9   Corporate Contributors  0 non-null      float64
 10  Former owner            1 non-null      object 
 11  Engraver                0 non-null      float64
 12  Issuance type           8287 non-null   object 
 13  Flickr URL              8287 non-null   object 
 14  Shelfmarks              8287 non-null   

## 3 Qu'est ce que vous remarquez?

> Dataset consiste de <b>8287 quartiers</b> et de <b>15 attributs</b> : 

>Identifier, Edition Statement, Place of Publication, Date of Publication, Date of Publication, Publisher,Title, Author, Contributors, Corporate Author, Corporate Contributors, Former owner, Engraver , Issuance type, Flickr URL ,Shelfmarks. 

> Il y'a aussi des données <b>manquantes</b> telle que on vérifier par:

><b>.isna().sum():</b> Compter le total des valeurs manquantes dans un dataframe.



In [7]:
missing = df.isna().sum()
print(missing)

Identifier                   0
Edition Statement         7514
Place of Publication         0
Date of Publication        181
Publisher                 4195
Title                        0
Author                    1778
Contributors                 0
Corporate Author          8287
Corporate Contributors    8287
Former owner              8286
Engraver                  8287
Issuance type                0
Flickr URL                   0
Shelfmarks                   0
dtype: int64


## 4 Dropping unnecessary columns

Nous allons restreindre l'étude à seulement 7 colonnes pour cela nous allons supprimer 7 colonnes que nous jugeons unitiles por le moment :
```Python
to_drop = ['Edition Statement',
           'Corporate Author',
           'Corporate Contributors',
           'Former owner',
           'Engraver',
           'Contributors',
           'Issuance type',
           'Shelfmarks']```

In [8]:
to_drop = ['Edition Statement',
           'Corporate Author',
           'Corporate Contributors',
           'Former owner',
           'Engraver',
           'Contributors',
           'Issuance type',
           'Shelfmarks']

df.drop(to_drop, inplace = True, axis = 1)
df.head()


Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                         Flickr URL  
0      A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
1  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
2  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
3  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...  
4  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...

## 5 Setting the index of the dataset
Changer l'index pourqu'ille prenne la valeur de l'identifiant `Identifier`

In [9]:
###Changerl'index 
df.set_index('Identifier', inplace = True)

###Afficher le résultat (sans le colonne index)
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

## 6 Donner un appérçu sur la colonne `Date of Publication` 


In [10]:
### Afficher  les 25 premières valeurs de Date of Publication
dop = df['Date of Publication']
dop.head(25)

Identifier
206            1879 [1878]
216                   1868
218                   1869
472                   1851
480                   1857
481                   1875
519                   1872
667                    NaN
874                   1676
1143                  1679
1280                  1802
1808                  1859
1905                  1888
1929           1839, 38-54
2836                  1897
2854                  1865
2956               1860-63
2957                  1873
3017                  1866
3131                  1899
4598                  1814
4884                  1820
4976                  1800
5382    1847, 48 [1846-48]
5385               [1897?]
Name: Date of Publication, dtype: object

> Remarque : Observer la date de publication qui correspond à l'index 206.
> Ca nécessite un nettoyage et une correction : il faut éliminer certains symboles 

## 7 Cleaning columns using the `.apply` function

### Ecrire une fonction qui permet de nettoyer la colonne des dates 

In [11]:
unwanted_characters = ['[', ',', '-']
# Completer la fonction suivante
def clean_dates(item):
    dop= str(item.loc['Date of Publication'])
    
    if dop == 'nan' or dop[0] == '[':
        return np.NAN
    
    for character in unwanted_characters:
        if character in dop:
            characters_index = dop.find(character)
            dop = dop[ :characters_index]
    
    return dop

df['Date of Publication'] = df.apply(clean_dates, axis = 1)

In [12]:
df.head(25)

Place of Publication Date of Publication  \
Identifier                                                            
206                                      London               1879    
216                    London; Virtue & Yorston                1868   
218                                      London                1869   
472                                      London                1851   
480                                      London                1857   
481                                      London                1875   
519                                      London                1872   
667         pp. 40. G. Bryan & Co: Oxford, 1898                 NaN   
874                                     London]                1676   
1143                                     London                1679   
1280                                   Coventry                1802   
1808                                Christiania                1859   
1905                                    Firenze                1888   
1929                                  Amsterdam                1839   
2836                                     Savona                1897   
2854                                     London                1865   
2956                                      Paris                1860   
2957                                      Paris                1873   
3017                                Puerto-Rico                1866   
3131                                   New York                1899   
4598                                       Hull                1814   
4884                                     London                1820   
4976                                     Oxonii                1800   
5382                                     London                1847   
5385                                     London                 NaN   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   
481             William Macintosh   
519                    The Author   
667                           NaN   
874                           NaN   
1143                          NaN   
1280         Printed by J. Turner   
1808                          NaN   
1905                          NaN   
1929                          NaN   
2836                          NaN   
2854               E. Moxon & Co.   
2956                          NaN   
2957                          NaN   
3017                          NaN   
3131                    W. Abbatt   
4598                   The Author   
4884            J. Hatchard & Son   
4976               J. Cooke, etc.   
5382                 Punch Office   
5385        Bradbury, Agnew & Co.   

                                                        Title  \
Identifier                                                      
206                         Walter Forbes. [A novel.] By A. A   
216         All for Greed. [A novel. The dedication signed...   
218         Love the Avenger. By the author of “All for Gr...   
472         Welsh Sketches, chiefly ecclesiastical, to the...   
480         [The World in which I live, and my place in it...   
481         [The World in which I live, and my place in it...   
519         Lagonells. By the author of Darmayne (F. E. A....   
667         The Coming of Spring, and other poems. By J. A...   
874         A Warning to the inhabitants of England, and L...   
1143        A Satyr against Vertue. (A poem: supposed to b...   
1280        An Account of the many and great Loans, Benefa...   
1808        Erindringer som Bidrag til Norges Historie fra...   
1905        Gli Studi storici in terra d'Otranto ... Framm...   
1929        De Aardbol. Magazijn van hedendaagsche land- e...   
2836        Cronache Savonesi dal 1500 al 1570 ... Accresc...   
2854

In [13]:
#Voici une autre méthode
#exécuter pour voir le résultat
unwanted_characters = ['[', ',', '-']

def clean_dates(dop):
    dop = str(dop)
    if dop.startswith('[') or dop == 'nan':
        return 'NaN'
    for character in unwanted_characters:
        if character in dop:
            character_index = dop.find(character)
            dop = dop[:character_index]
    return dop

df['Date of Publication'] = df['Date of Publication'].apply(clean_dates)
df.head(10)

Place of Publication Date of Publication  \
Identifier                                                            
206                                      London               1879    
216                    London; Virtue & Yorston                1868   
218                                      London                1869   
472                                      London                1851   
480                                      London                1857   
481                                      London                1875   
519                                      London                1872   
667         pp. 40. G. Bryan & Co: Oxford, 1898                 NaN   
874                                     London]                1676   
1143                                     London                1679   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   
481             William Macintosh   
519                    The Author   
667                           NaN   
874                           NaN   
1143                          NaN   

                                                        Title         Author  \
Identifier                                                                     
206                         Walter Forbes. [A novel.] By A. A          A. A.   
216         All for Greed. [A novel. The dedication signed...      A., A. A.   
218         Love the Avenger. By the author of “All for Gr...      A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...      A., E. S.   
480         [The World in which I live, and my place in it...      A., E. S.   
481         [The World in which I live, and my place in it...      A., E. S.   
519         Lagonells. By the author of Darmayne (F. E. A....      A., F. E.   
667         The Coming of Spring, and other poems. By J. A...  A., J.|A., J.   
874         A Warning to the inhabitants of England, and L...         Remaʿ.   
1143        A Satyr against Vertue. (A poem: supposed to b...         A., T.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...  
481         http://www.flickr.com/photos/britishlibrary/ta...  
519         http://www.flickr.com/photos/britishlibrary/ta...  
667         http://www.flickr.com/photos/britishlibrary/ta...  
874         http://www.flickr.com/photos/britishlibrary/ta...  
1143        http://www.flickr.com/photos/britishlibrary/ta...

## 8 Observer la colonne `Author`
### Completer la fonction `clean_author_names`qui permet de corriger les valeurs de cet colonne

In [14]:
# compléter la fonction suivante ( à la place de pass)
def clean_author_names(author):
    
    author = str(author)
    
    if author == 'nan':
        return 'NaN'
    
    author = author.split(',')

    if len(author) == 1:
        name = filter(lambda x: x.isalpha(), author[0])
        return reduce(lambda x, y: x + y, name)
    
    last_name, first_name = author[0], author[1]

    first_name = first_name[:first_name.find('-')] if '-' in first_name else first_name
    
    if first_name.endswith(('.', '.|')):
        parts = first_name.split('.')
        
        if len(parts) > 1:
            first_occurence = first_name.find('.')
            final_occurence = first_name.find('.', first_occurence + 1)
            first_name = first_name[:final_occurence]
            
        else:
            first_name = first_name[:first_name.find('.')]
    
    last_name = last_name.capitalize()
    return f'{first_name} {last_name}'

df['Author'] = df['Author'].apply(clean_author_names)
df.head(10)

Place of Publication Date of Publication  \
Identifier                                                            
206                                      London               1879    
216                    London; Virtue & Yorston                1868   
218                                      London                1869   
472                                      London                1851   
480                                      London                1857   
481                                      London                1875   
519                                      London                1872   
667         pp. 40. G. Bryan & Co: Oxford, 1898                 NaN   
874                                     London]                1676   
1143                                     London                1679   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   
481             William Macintosh   
519                    The Author   
667                           NaN   
874                           NaN   
1143                          NaN   

                                                        Title    Author  \
Identifier                                                                
206                         Walter Forbes. [A novel.] By A. A        AA   
216         All for Greed. [A novel. The dedication signed...   A. A A.   
218         Love the Avenger. By the author of “All for Gr...   A. A A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...   E. S A.   
480         [The World in which I live, and my place in it...   E. S A.   
481         [The World in which I live, and my place in it...   E. S A.   
519         Lagonells. By the author of Darmayne (F. E. A....   F. E A.   
667         The Coming of Spring, and other poems. By J. A...   J.|A A.   
874         A Warning to the inhabitants of England, and L...     Remaʿ   
1143        A Satyr against Vertue. (A poem: supposed to b...      T A.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...  
481         http://www.flickr.com/photos/britishlibrary/ta...  
519         http://www.flickr.com/photos/britishlibrary/ta...  
667         http://www.flickr.com/photos/britishlibrary/ta...  
874         http://www.flickr.com/photos/britishlibrary/ta...  
1143        http://www.flickr.com/photos/britishlibrary/ta...

# Résultat attendu 
<img src="Datasets\capture42.png">

In [15]:
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London               1879    
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title    Author  \
Identifier                                                                
206                         Walter Forbes. [A novel.] By A. A        AA   
216         All for Greed. [A novel. The dedication signed...   A. A A.   
218         Love the Avenger. By the author of “All for Gr...   A. A A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...   E. S A.   
480         [The World in which I live, and my place in it...   E. S A.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

##  Observer la colonne `title`
### 9 Completer la fonction `clean_title`qui permet de corriger les valeurs de cet colonne

# Résultat attendu 
 <img src="Datasets/capture43.png">

In [16]:
## complétrer la fonction suivante ( à la place de pass)

def clean_title(title):
    
    if title == 'nan':
        return 'NaN'
    
    if title[0] == '[':
        title = title[1: title.find(']')]
        
    if 'by' in title:
        title = title[:title.find('by')]
    elif 'By' in title:
        title = title[:title.find('By')]
        
    if '[' in title:
        title = title[:title.find('[')]

    title = title[:-2]
        
    title = list(map(str.capitalize, title.split()))
    return ' '.join(title)
    
df['Title'] = df['Title'].apply(clean_title)
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London               1879    
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title    Author  \
Identifier                                                                
206                                             Walter Forbes        AA   
216                                             All For Greed   A. A A.   
218                                          Love The Avenger   A. A A.   
472         Welsh Sketches, Chiefly Ecclesiastical, To The...   E. S A.   
480             The World In Which I Live, And My Place In It   E. S A.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

# PART II
## 1 Cleaning entire dataset
### On considère le fichier suivant 'university_towns.txt'

In [17]:
!head Datasets/university_towns.txt

'head' n'est pas reconnu en tant que commande interne
ou externe, un programme ex‚cutable ou un fichier de commandes.


In [18]:
with open('Datasets/university_towns.txt', 'r') as text:
    textfile = text.read()
    print(textfile)

Alabama[edit]
Auburn (Auburn University)[1]
Florence (University of North Alabama)
Jacksonville (Jacksonville State University)[2]
Livingston (University of West Alabama)[2]
Montevallo (University of Montevallo)[2]
Troy (Troy University)[2]
Tuscaloosa (University of Alabama, Stillman College, Shelton State)[3][4]
Tuskegee (Tuskegee University)[5]
Alaska[edit]
Fairbanks (University of Alaska Fairbanks)[2]
Arizona[edit]
Flagstaff (Northern Arizona University)[6]
Tempe (Arizona State University)
Tucson (University of Arizona)
Arkansas[edit]
Arkadelphia (Henderson State University, Ouachita Baptist University)[2]
Conway (Central Baptist College, Hendrix College, University of Central Arkansas)[2]
Fayetteville (University of Arkansas)[7]
Jonesboro (Arkansas State University)[8]
Magnolia (Southern Arkansas University)[2]
Monticello (University of Arkansas at Monticello)[2]
Russellville (Arkansas Tech University)[2]
Searcy (Harding University)[5]
California[edit]
Angwin (Pacific Union College

## 2 Créer un pandas data frame à partir d'un fichier txt
On veut créer un DataFrame nommé `towns_df` à partir du fichier 'university_towns.txt'

Nous voyons que nous avons des noms d'états périodiques suivis des villes universitaires de cet état: StateA TownA1 TownA2 StateB TownB1 TownB2 .... Si nous regardons la façon dont les noms d'états sont écrits dans le fichier, nous verrons qu'ils ont tous la sous-chaîne «[edit]» en eux.

Nous pouvons tirer parti de ce modèle en créant une liste de tuples (état, ville) et en enveloppant cette liste dans un DataFrame:


In [19]:
university_towns = []
with open('Datasets/university_towns.txt') as file:
    for line in file:
        if '[edit]' in line:
             # se rappeler de ce `state` jusqu'à ce qu'on trouve le suivant
             state= line
        else:
             # Sinon, nous avons une ville; garder `state` 
             university_towns.append((state,line))

university_towns[:5]

[('Alabama[edit]\n', 'Auburn (Auburn University)[1]\n'),
 ('Alabama[edit]\n', 'Florence (University of North Alabama)\n'),
 ('Alabama[edit]\n', 'Jacksonville (Jacksonville State University)[2]\n'),
 ('Alabama[edit]\n', 'Livingston (University of West Alabama)[2]\n'),
 ('Alabama[edit]\n', 'Montevallo (University of Montevallo)[2]\n')]

In [20]:
towns_df = pd.DataFrame(university_towns,
                        columns=['State', 'RegionName'])
towns_df.head()

State                                         RegionName
0  Alabama[edit]\n                    Auburn (Auburn University)[1]\n
1  Alabama[edit]\n           Florence (University of North Alabama)\n
2  Alabama[edit]\n  Jacksonville (Jacksonville State University)[2]\n
3  Alabama[edit]\n       Livingston (University of West Alabama)[2]\n
4  Alabama[edit]\n         Montevallo (University of Montevallo)[2]\n

## 3 Compléter la fonction `clean_up` pour nettoyer les colonnes de `towns_df`

In [21]:
## complétrer la fonction suivante ( à la place de pass)
def clean_up(item):
    if '(' in item:
        pass
        return item[:item.find('(') - 1]
    
    if '[' in item:
        return item[:item.find('[')]
    

towns_df =  towns_df.applymap(clean_up)
towns_df

State     RegionName
0      Alabama         Auburn
1      Alabama       Florence
2      Alabama   Jacksonville
3      Alabama     Livingston
4      Alabama     Montevallo
..         ...            ...
512  Wisconsin    River Falls
513  Wisconsin  Stevens Point
514  Wisconsin       Waukesha
515  Wisconsin     Whitewater
516    Wyoming        Laramie

[517 rows x 2 columns]

> <b>applymap</b> : Appliquer une fonction à un élément Dataframe. Cette méthode applique une fonction qui accepte et renvoie un scalaire à chaque élément d'un DataFrame.

# PART III ( OPTIONNELLE) 
## 1 Renaming columns and skipping rows
### 1.1 Convertir le fichier `olympics.csv` en dataframe pandas `olympics_df`

In [22]:
# exporter 'olympics.csv' sous forme de data frame nommé olympics_df
olympics_df = pd.read_csv('Datasets/olympics.csv')

# Afficher les 5 premières lignes
olympics_df.head()

0         1     2     3     4      5         6     7     8  \
0                NaN  ? Summer  01 !  02 !  03 !  Total  ? Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  ? Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

### 1.2 Supprimer la première ligne 

In [23]:
# exporter 'olympics.csv' sous forme de data frame nommé olympics_df en supprimant la première ligne

olympics_df = pd.read_csv('./Datasets/olympics.csv', skiprows = 1, header = 0)

# Afficher les 5 premières lignes
olympics_df.head()

Unnamed: 0  ? Summer  01 !  02 !  03 !  Total  ? Winter  \
0        Afghanistan (AFG)        13     0     0     2      2         0   
1            Algeria (ALG)        12     5     2     8     15         3   
2          Argentina (ARG)        23    18    24    28     70        18   
3            Armenia (ARM)         5     1     2     9     12         6   
4  Australasia (ANZ) [ANZ]         2     3     4     5     12         0   

   01 !.1  02 !.1  03 !.1  Total.1  ? Games  01 !.2  02 !.2  03 !.2  \
0       0       0       0        0       13       0       0       2   
1       0       0       0        0       15       5       2       8   
2       0       0       0        0       41      18      24      28   
3       0       0       0        0       11       1       2       9   
4       0       0       0        0        2       3       4       5   

   Combined total  
0               2  
1              15  
2              70  
3              12  
4              12

### 1.3 Changer le nom des colonnes en utilisant le dictionnaire suivant :
```python
new_names =  {'Unnamed: 0': 'Country',
              '? Summer': 'Summer Olympics',
              '01 !': 'Gold',
              '02 !': 'Silver',
              '03 !': 'Bronze',
              '? Winter': 'Winter Olympics',
              '01 !.1': 'Gold.1',
              '02 !.1': 'Silver.1',
              '03 !.1': 'Bronze.1',
              '? Games': '# Games', 
              '01 !.2': 'Gold.2',
              '02 !.2': 'Silver.2',
              '03 !.2': 'Bronze.2'}


```



In [24]:
# Changer le nom des colonnes en utilisant le dictionnaire new_names
new_names =  {'Unnamed: 0': 'Country',
              '? Summer': 'Summer Olympics',
              '01 !': 'Gold',
              '02 !': 'Silver',
              '03 !': 'Bronze',
              '? Winter': 'Winter Olympics',
              '01 !.1': 'Gold.1',
              '02 !.1': 'Silver.1',
              '03 !.1': 'Bronze.1',
              '? Games': '# Games', 
              '01 !.2': 'Gold.2',
              '02 !.2': 'Silver.2',
              '03 !.2': 'Bronze.2'}

olympics_df.rename(columns = new_names, inplace = True)

# Afficher les 5 premières lignes
olympics_df.head()

Country  Summer Olympics  Gold  Silver  Bronze  Total  \
0        Afghanistan (AFG)               13     0       0       2      2   
1            Algeria (ALG)               12     5       2       8     15   
2          Argentina (ARG)               23    18      24      28     70   
3            Armenia (ARM)                5     1       2       9     12   
4  Australasia (ANZ) [ANZ]                2     3       4       5     12   

   Winter Olympics  Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
0                0       0         0         0        0       13       0   
1                3       0         0         0        0       15       5   
2               18       0         0         0        0       41      18   
3                6       0         0         0        0       11       1   
4                0       0         0         0        0        2       3   

   Silver.2  Bronze.2  Combined total  
0         0         2               2  
1         2         8              15  
2        24        28              70  
3         2         9              12  
4         4         5              12

In [25]:
olympics_df.head()

Country  Summer Olympics  Gold  Silver  Bronze  Total  \
0        Afghanistan (AFG)               13     0       0       2      2   
1            Algeria (ALG)               12     5       2       8     15   
2          Argentina (ARG)               23    18      24      28     70   
3            Armenia (ARM)                5     1       2       9     12   
4  Australasia (ANZ) [ANZ]                2     3       4       5     12   

   Winter Olympics  Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
0                0       0         0         0        0       13       0   
1                3       0         0         0        0       15       5   
2               18       0         0         0        0       41      18   
3                6       0         0         0        0       11       1   
4                0       0         0         0        0        2       3   

   Silver.2  Bronze.2  Combined total  
0         0         2               2  
1         2         8              15  
2        24        28              70  
3         2         9              12  
4         4         5              12

> <b>On fait cleanning data sur deux types de données l'un de type csv et l'autre txt, toute en faisant les étapes d'exploration ainsi que suppression unnecessary columns et les cleanning...</b>

In [1]:
import session_info
session_info.show()